sw/device/lib/testing/test_framework/ottf_flow_control.c

```diff
@@ -0,0 +1,121 @@
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_isrs.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static flow_control_t flow_control_state;
+#define FLOW_CONTROL_LOW_WATERMARK 4
+#define FLOW_CONTROL_HIGH_WATERMARK 16
+
+const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+dif_rv_plic_t ottf_plic;
+uint32_t ottf_flow_control_intr;
+
+void ottf_flow_control_enable(void) {
+  CHECK_DIF_OK(dif_rv_plic_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR), &ottf_plic));
+
+  dif_uart_t *uart = ottf_console();
+  CHECK_DIF_OK(dif_uart_watermark_rx_set(uart, kDifUartWatermarkByte16));
+  CHECK_DIF_OK(dif_uart_irq_set_enabled(uart, kDifUartIrqRxWatermark,
+                                        kDifToggleEnabled));
+
+  // Set IRQ priorities to MAX
+  CHECK_DIF_OK(dif_rv_plic_irq_set_priority(
+      &ottf_plic, kTopEarlgreyPlicIrqIdUart0RxWatermark,
+      kDifRvPlicMaxPriority));
+  // Set Ibex IRQ priority threshold level
+  CHECK_DIF_OK(dif_rv_plic_target_set_threshold(&ottf_plic, kPlicTarget,
+                                                kDifRvPlicMinPriority));
+  // Enable IRQs in PLIC
+  CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
+      &ottf_plic, kTopEarlgreyPlicIrqIdUart0RxWatermark, kPlicTarget,
+      kDifToggleEnabled));
+
+  flow_control_state = kFlowControlAuto;
+  irq_global_ctrl(true);
+  irq_external_ctrl(true);
+}
+
+// This version of the function is safe to call from within the ISR.
+static status_t manage_flow_control(const dif_uart_t *uart,
+                                    flow_control_t ctrl) {
+  if (flow_control_state == kFlowControlNone) {
+    return OK_STATUS(flow_control_state);
+  }
+  if (ctrl == kFlowControlAuto) {
+    uint32_t avail;
+    TRY(dif_uart_rx_bytes_available(uart, &avail));
+    if (avail < FLOW_CONTROL_LOW_WATERMARK &&
+        flow_control_state != kFlowControlResume) {
+      ctrl = kFlowControlResume;
+    } else if (avail >= FLOW_CONTROL_HIGH_WATERMARK &&
+               flow_control_state != kFlowControlPause) {
+      ctrl = kFlowControlPause;
+    } else {
+      return OK_STATUS(flow_control_state);
+    }
+  }
+  uint8_t byte = (uint8_t)ctrl;
+  CHECK_DIF_OK(dif_uart_bytes_send(uart, &byte, 1, NULL));
+  flow_control_state = ctrl;
+  return OK_STATUS(flow_control_state);
+}
+
+bool ottf_flow_control_isr(void) {
+  ottf_flow_control_intr += 1;
+  dif_uart_t *uart = ottf_console();
+  bool rx;
+  CHECK_DIF_OK(dif_uart_irq_is_pending(uart, kDifUartIrqRxWatermark, &rx));
+  if (rx) {
+    manage_flow_control(uart, kFlowControlPause);
+    CHECK_DIF_OK(dif_uart_irq_acknowledge(uart, kDifUartIrqRxWatermark));
+    return true;
+  }
+  return false;
+}
+
+void ottf_external_isr(void) {
```

Comment:
As-is, this might need to be booted out of the framework and down to the test level, as it prevents any other (external) interrupt handlers from being made available (and might cause link errors if others use it?). WDYT?

Some ideas:
1. Rename it and create a function that we require every integrator to call in their top-level ISR.
2. Create an ISR registration system in OTTF that handles dispatch.

Reply:
I'm leaning towards option 1: in your ISR, call `ottf_flow_control_isr()`.

I'll probably have to refactor the plic init to be in `ottf_isrs` and adjust `ottf_flow_control_isr()` to do all the appropriate plic checking and clearing.